### PR TITLE
Add quadratic discriminant analysis classifier

### LINKED ISSUE #338: Add Linear and Quadratic Discriminant Analysis (LDA / QDA)
Noticed Scholar doesn't have discriminant analysis yet, would be nice to add:

- [x] LinearDiscriminantAnalysis (LDA)
- [ ] QuadraticDiscriminantAnalysis (QDA)

Refs:
- https://scikit-learn.org/stable/modules/lda_qda.html
- https://scikit-learn.org/stable/modules/generated/sklearn.discriminant_analysis.LinearDiscriminantAnalysis.html
- https://scikit-learn.org/stable/modules/generated/sklearn.discriminant_analysis.QuadraticDiscriminantAnalysis.html

### PR DESCRIPTION
**Pull request by RicardoSantos-99**
Closes #338. LDA landed in #341, so this completes it.

Adds `Scholar.DiscriminantAnalysis.Quadratic`. Each class keeps its own covariance, stored as its eigendecomposition so both the log determinant and the quadratic form become elementwise work over the eigenvalues. Options are `:num_classes` (required, as in LDA) and `:reg_param`.

### Two implementation notes

Covariances are built and decomposed one class at a time. That keeps the working set at O(num_samples * num_features), and it avoids `Nx.LinAlg.eigh/2` over a stack of matrices, which on Nx 0.9.x under EXLA decomposes only the first one and returns zeros for the rest. Fixed in Nx 0.13, but Scholar still allows `~> 0.9`.

`eigh` is called with a tighter `eps` than its default. QDA divides by the eigenvalues and takes their log, and the default `1.0e-4` leaves them about three digits short in the pure Nx implementation. Compiled backends replace that routine and are unaffected.

### Validation against scikit-learn 1.6.1

Twelve scenarios (2 to 5 classes, up to 10 features, unbalanced, overlapping, correlated features, mixed scales, `reg_param`, singular covariance), EXLA in f64: means, eigenvalues and scores within 6.0e-8, identical predictions in eleven of twelve.

On a 30% stratified holdout, every prediction agrees:

| dataset | this PR | scikit-learn | LDA |
|---|---|---|---|
| iris | 1.0000 | 1.0000 | 1.0000 |
| wine | 0.9815 | 0.9815 | 1.0000 |
| breast_cancer | 0.9532 | 0.9532 | 0.9532 |
| digits (61 features, 10 classes) | 0.9704 | 0.9704 | 0.9593 |

### One deliberate divergence

The twelfth scenario is a class with fewer samples than features. scikit-learn keeps only `min(class_size, num_features)` components and works in the subspace the class spans; this keeps all directions and relies on `:reg_param`. Noted in the moduledoc.

Tests include the `jit_apply` case AGENTS.md asks for, and two classes that share a mean and differ only in spread, where LDA cannot separate them.

### CODE REVIEW

**Comment by RicardoSantos-99** [comment]
@josevalim while working on this I hit a bug in the Nx the lockfile pins. On 0.9.2, `Nx.LinAlg.eigh/2` over a stack of matrices decomposes only the first one under EXLA and returns zeros for the rest:

```elixir
one = Nx.tensor([[2.0, 1.0], [1.0, 2.0]])

Nx.Defn.jit_apply(
  fn t -> Nx.LinAlg.eigh(t) |> elem(0) end,
  [Nx.stack([one, one])],
  compiler: EXLA
)
#=> [[1.0, 3.0], [0.0, 0.0]]
``` 

It is nondeterministic, sometimes exiting the VM silently instead of returning. Both routes into EXLA's native routine are affected, the compiler and EXLA.Backend on its own. BinaryBackend is fine, and it is already fixed in Nx 0.13, which I confirmed on a clean project.

Worth bumping? It fails silently with wrong numbers rather than raising, so anything in Scholar reaching a batched decomposition is exposed, not just my case. I worked around it by decomposing one class at a time, but that workaround only exists because of 0.9.

**Comment by josevalim** [comment]
> Fixed in Nx 0.13, but Scholar still allows ~> 0.9.

We can bump the requirement to v0.13 if that is a better solution.

File: lib/scholar/discriminant_analysis/quadratic.ex [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+  end
+
+  defnp fit_n(x, y, opts) do
+    x = to_float(x)
+    num_classes = opts[:num_classes]
+    reg_param = opts[:reg_param]
+    {num_samples, num_features} = Nx.shape(x)
+
+    one_hot =
+      (Nx.new_axis(y, 1) == Nx.new_axis(Nx.iota({num_classes}), 0)) |> Nx.as_type(Nx.type(x))
```
Comment (josevalim) [inline lib/scholar/discriminant_analysis/quadratic.ex:141]: Don't we already have a function that does this somewhere in the lib?

Comment (RicardoSantos-99) [reply lib/scholar/discriminant_analysis/quadratic.ex:141]: I think the `Scholar.Preprocessing.one_hot_encode/2` is the closest one, but it runs the labels through `OrdinalEncoder` first, which compacts them to the distinct values actually present. With `num_classes: 3`:

| y | inline | `one_hot_encode` |
|---|---|---|
| `[0, 2, 2, 0]` | counts `[2, 0, 2]` | counts `[2, 2, 0]` |
| `[1, 1, 2, 2]` | counts `[0, 2, 2]` | counts `[2, 2, 0]` |

So a class missing from the training data shifts every class after it, and the means, priors and covariances would silently line up with the wrong class. That conflicts with the contract documented for `:num_classes` here, which is that labels are already integers in `[0, num_classes)`.

Comment (RicardoSantos-99) [reply lib/scholar/discriminant_analysis/quadratic.ex:141]: But the line is duplicated, because `DiscriminantAnalysis.Linear` has the same one, and `LogisticRegression`, `NaiveBayes.Complement` and `Metrics.Classification` have it in a slightly different spelling. 

@josevalim do you think it's better extract to `Scholar.Shared` and use it here and in LDA?, leaving the other three for a follow up.

Comment (josevalim) [reply lib/scholar/discriminant_analysis/quadratic.ex:141]: Nah, it is fine to have it duplicated, it is a single line of code anyway!

**Comment by josevalim** [comment]
Btw, if we are bumping Nx, let's do that on another PR and then we rebase this one!

**Comment by RicardoSantos-99** [comment]
> Btw, if we are bumping Nx, let's do that on another PR and then we rebase this one!

@josevalim I tried the bump locally to see how big it is, and it is not a small change.

Two things came up. Nx and EXLA 0.13 declare elixir: `"~> 1.17"`, so Scholar's own requirement moves off `~> 1.14` and the older leg of the CI matrix moves off 1.15.8. That drops 1.14 through 1.16 in one go

The suite also goes to 157 failures on 0.13. Most of it is harmless: 139 are doctests, and 0.13 prints `f32` with the shortest form that round trips, so `7.071067810058594` is now `7.071068`. The rest needs judgment: around 26 cases have values that actually moved, in TSNE, TruncatedSVD, PLSSVD, etc.

For this PR the bump changes nothing structurally. The only thing it buys is dropping the `@eigh_eps` constant, since on 0.13 passing `:eps` makes no difference and the default is already accurate. So I would rather not hold this one on it. Happy to pick the bump up separately whenever you want, and it probably wants to be two PRs, the version plus the formatting churn first, then the fixtures that moved.

**Review by josevalim — APPROVED** [review]

**Comment by josevalim** [comment]
Approved this one! A PR for the upgrade would be fantastic as well, thanks!

**Comment by josevalim** [comment]
:green_heart: :blue_heart: :purple_heart: :yellow_heart: :heart:
